bser.c

```diff
@@ -0,0 +1,163 @@
+#include <string.h>
+#include <stdarg.h>
+#include <assert.h>
+
+#include "bser.h"
+#include "bser_private.h"
+
+bser_t* bser_new_integer(uint64_t value, bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_INT64;
+    fill->value.integer = value;
+    return fill;
+}
+
+bser_t* bser_new_real(double value, bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_REAL;
+    fill->value.real = value;
+    return fill;
+}
+
+bser_t* bser_new_true(bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_TRUE;
+    return fill;
+}
+
+bser_t* bser_new_false(bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_FALSE;
+    return fill;
+}
+
+bser_t* bser_new_null(bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_NULL;
+    return fill;
+}
+
+bser_t* bser_new_string(const char* chars, size_t len, bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_STRING;
+    fill->value.string.chars = chars;
+    fill->value.string.length = len;
+    return fill;
+}
+
+bser_t* bser_new_array(
+        bser_t* elems, size_t len, bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_ARRAY;
+    fill->value.array.elements = elems;
+    fill->value.array.length = len;
+    return fill;
+}
+
+bser_t* bser_new_object(
+        bser_key_value_pair_t* fields, size_t length, bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_OBJECT;
+    fill->value.object.fields = fields;
+    fill->value.object.length = length;
+    return fill;
+}
+
+static void bser_free_array(bser_t* bser) {
```

Comment:
This does not actually do what the name says it will do; it will free the array's elements but not the array itself.